File: test/language/array_types/cpp/VariableArrayStructCastVarUIntTest.cpp
#include "array_types/variable_array_struct_cast_varuint/VariableArray.h"
#include "gtest/gtest.h"
#include "zserio/RebindAlloc.h"
#include "zserio/SerializeUtil.h"
#include "zserio/StringConvertUtil.h"

namespace array_types
{
namespace variable_array_struct_cast_varuint
{

using allocator_type = VariableArray::allocator_type;
using string_type = zserio::string<allocator_type>;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class VariableArrayStructCastVarUIntTest : public ::testing::Test
{
protected:
    void writeVariableArrayToByteArray(zserio::BitStreamWriter& writer, size_t numElements)
    {
        writer.writeBits64(static_cast<uint64_t>(numElements), 8);
        for (size_t i = 0; i < numElements; ++i)
        {
            writer.writeBits(static_cast<uint32_t>(i), 32);
            writer.writeString(std::string("Name") + std::to_string(i));
        }
    }

    static const std::string BLOB_NAME;
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const std::string VariableArrayStructCastVarUIntTest::BLOB_NAME =
        "language/array_types/variable_array_struct_cast_varuint.blob";

TEST_F(VariableArrayStructCastVarUIntTest, bitSizeOf)
{
    const size_t numElements = 33;
    vector_type<TestStructure> compoundArray;
    compoundArray.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
    {
        TestStructure testStructure;
        testStructure.setId(static_cast<uint32_t>(i));
        testStructure.setName(string_type("Name") + zserio::toString<allocator_type>(i));
        compoundArray.push_back(testStructure);
    }
    VariableArray variableArray;
    variableArray.setNumElements(static_cast<uint64_t>(numElements));
    variableArray.setCompoundArray(compoundArray);

    const size_t bitPosition = 2;
    const size_t numOneNumberIndexes = 10; // (numElements > 9) ? 10 : numElements
    const size_t expectedBitSize = (1 + numElements * (4 + 7) - numOneNumberIndexes) * 8;
    ASSERT_EQ(expectedBitSize, variableArray.bitSizeOf(bitPosition));
}

TEST_F(VariableArrayStructCastVarUIntTest, initializeOffsets)
{
    const size_t numElements = 33;
    vector_type<TestStructure> compoundArray;
    compoundArray.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
    {
        TestStructure testStructure;
        testStructure.setId(static_cast<uint32_t>(i));
        testStructure.setName(string_type("Name") + zserio::toString<allocator_type>(i));
        compoundArray.push_back(testStructure);
    }
    VariableArray variableArray;
    variableArray.setNumElements(static_cast<uint64_t>(numElements));
    variableArray.setCompoundArray(compoundArray);

    const size_t bitPosition = 2;
    const size_t numOneNumberIndexes = 10; // (numElements > 9) ? 10 : numElements
    const size_t expectedEndBitPosition = bitPosition + (1 + numElements * (4 + 7) - numOneNumberIndexes) * 8;
    ASSERT_EQ(expectedEndBitPosition, variableArray.initializeOffsets(bitPosition));
}

TEST_F(VariableArrayStructCastVarUIntTest, readConstructor)
{
    const size_t numElements = 59;
    zserio::BitStreamWriter writer(bitBuffer);
    writeVariableArrayToByteArray(writer, numElements);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    VariableArray variableArray(reader);

    ASSERT_EQ(numElements, static_cast<size_t>(variableArray.getNumElements()));
    const vector_type<TestStructure>& compoundArray = variableArray.getCompoundArray();
    ASSERT_EQ(numElements, compoundArray.size());
    for (size_t i = 0; i < numElements; ++i)
    {
        ASSERT_EQ(i, compoundArray[i].getId());
        ASSERT_EQ(std::string("Name") + std::to_string(i), compoundArray[i].getName().c_str());
    }
}

TEST_F(VariableArrayStructCastVarUIntTest, writeRead)
{
    const size_t numElements = 33;
    vector_type<TestStructure> compoundArray;
    compoundArray.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
    {
        TestStructure testStructure;
        testStructure.setId(static_cast<uint32_t>(i));
        testStructure.setName(string_type("Name") + zserio::toString<allocator_type>(i));
        compoundArray.push_back(testStructure);
    }
    VariableArray variableArray;
    variableArray.setNumElements(static_cast<uint64_t>(numElements));
    variableArray.setCompoundArray(compoundArray);

    zserio::BitStreamWriter writer(bitBuffer);
    variableArray.write(writer);

    ASSERT_EQ(variableArray.bitSizeOf(), writer.getBitPosition());
    ASSERT_EQ(variableArray.initializeOffsets(), writer.getBitPosition());

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    VariableArray readVariableArray(reader);
    const vector_type<TestStructure>& readCompoundArray = readVariableArray.getCompoundArray();
    ASSERT_EQ(numElements, readCompoundArray.size());
    for (size_t i = 0; i < numElements; ++i)
    {
        ASSERT_EQ(i, readCompoundArray[i].getId());
        ASSERT_EQ(std::string("Name") + std::to_string(i), readCompoundArray[i].getName().c_str());
    }
}

TEST_F(VariableArrayStructCastVarUIntTest, writeReadFile)
{
    const size_t numElements = 33;
    vector_type<TestStructure> compoundArray;
    compoundArray.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
    {
        TestStructure testStructure;
        testStructure.setId(static_cast<uint32_t>(i));
        testStructure.setName(string_type("Name") + zserio::toString<allocator_type>(i));
        compoundArray.push_back(testStructure);
    }
    VariableArray variableArray;
    variableArray.setNumElements(static_cast<uint64_t>(numElements));
    variableArray.setCompoundArray(compoundArray);

    zserio::serializeToFile(variableArray, BLOB_NAME);

    VariableArray readVariableArray = zserio::deserializeFromFile<VariableArray>(BLOB_NAME);
    const vector_type<TestStructure>& readCompoundArray = readVariableArray.getCompoundArray();
    ASSERT_EQ(numElements, readCompoundArray.size());
    for (size_t i = 0; i < numElements; ++i)
    {
        ASSERT_EQ(i, readCompoundArray[i].getId());
        ASSERT_EQ(std::string("Name") + std::to_string(i), readCompoundArray[i].getName().c_str());
    }
}

TEST_F(VariableArrayStructCastVarUIntTest, writeWrongArray)
{
    const size_t numElements = 33;
    vector_type<TestStructure> compoundArray;
    compoundArray.reserve(numElements);
    for (size_t i = 0; i < numElements; ++i)
    {
        TestStructure testStructure;
        testStructure.setId(static_cast<uint32_t>(i));
        testStructure.setName(string_type("Name") + zserio::toString<allocator_type>(i));
        compoundArray.push_back(testStructure);
    }
    VariableArray variableArray;
    variableArray.setNumElements(static_cast<uint64_t>(numElements) + 1);
    variableArray.setCompoundArray(compoundArray);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(variableArray.write(writer), zserio::CppRuntimeException);
}

} // namespace variable_array_struct_cast_varuint
} // namespace array_types

File: test/extensions/language/array_types/java/array_types/VariableArrayStructCastVarUIntTest.java
package array_types;

import static org.junit.jupiter.api.Assertions.*;

import java.io.File;
import java.io.IOException;
import java.math.BigInteger;

import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

import array_types.variable_array_struct_cast_varuint.TestStructure;
import array_types.variable_array_struct_cast_varuint.VariableArray;

public class VariableArrayStructCastVarUIntTest
{
    @Test
    public void bitSizeOf() throws IOException, ZserioError
    {
        final byte numElements = 33;
        final TestStructure[] compoundArray = new TestStructure[numElements];
        for (byte i = 0; i < numElements; ++i)
        {
            final TestStructure testStructure = new TestStructure(i, "Name" + i);
            compoundArray[i] = testStructure;
        }
        final VariableArray variableArray = new VariableArray(BigInteger.valueOf(numElements), compoundArray);
        final int bitPosition = 2;
        final int numOneNumberIndexes = 10;
        final int expectedBitSize = (1 + numElements * (4 + 7) - numOneNumberIndexes) * 8;
        assertEquals(expectedBitSize, variableArray.bitSizeOf(bitPosition));
    }

    @Test
    public void initializeOffsets() throws IOException, ZserioError
    {
        final byte numElements = 33;
        final TestStructure[] compoundArray = new TestStructure[numElements];
        for (byte i = 0; i < numElements; ++i)
        {
            final TestStructure testStructure = new TestStructure(i, "Name" + i);
            compoundArray[i] = testStructure;
        }
        final VariableArray variableArray = new VariableArray(BigInteger.valueOf(numElements), compoundArray);
        final int bitPosition = 2;
        final int numOneNumberIndexes = 10;
        final int expectedEndBitPosition = bitPosition + (1 + numElements * (4 + 7) - numOneNumberIndexes) * 8;
        assertEquals(expectedEndBitPosition, variableArray.initializeOffsets(bitPosition));
    }

    @Test
    public void read() throws IOException, ZserioError
    {
        final byte numElements = 59;
        final BitBuffer buffer = writeVariableArrayToBitBuffer(numElements);
        final VariableArray variableArray = SerializeUtil.deserialize(VariableArray.class, buffer);

        assertEquals(BigInteger.valueOf(numElements), variableArray.getNumElements());
        final TestStructure[] compoundArray = variableArray.getCompoundArray();
        assertEquals(numElements, compoundArray.length);
        for (byte i = 0; i < numElements; ++i)
        {
            final TestStructure testStructure = compoundArray[i];
            assertEquals(i, testStructure.getId());
            assertTrue(testStructure.getName().equals("Name" + i));
        }
    }

    @Test
    public void writeRead() throws IOException, ZserioError
    {
        final byte numElements = 33;
        final TestStructure[] compoundArray = new TestStructure[numElements];
        for (short i = 0; i < numElements; ++i)
        {
            final TestStructure testStructure = new TestStructure(i, "Name" + i);
            compoundArray[i] = testStructure;
        }
        final VariableArray variableArray = new VariableArray(BigInteger.valueOf(numElements), compoundArray);
        final File file = new File(BLOB_NAME);
        SerializeUtil.serializeToFile(variableArray, file);

        final VariableArray readVariableArray = SerializeUtil.deserializeFromFile(VariableArray.class, file);
        assertEquals(BigInteger.valueOf(numElements), readVariableArray.getNumElements());
        final TestStructure[] readCompoundArray = readVariableArray.getCompoundArray();
        assertEquals(numElements, readCompoundArray.length);
        for (byte i = 0; i < numElements; ++i)
        {
            final TestStructure readTestStructure = readCompoundArray[i];
            assertEquals(i, readTestStructure.getId());
            assertTrue(readTestStructure.getName().equals("Name" + i));
        }
    }

    @Test
    public void writeWrongArray() throws IOException, ZserioError
    {
        final byte numElements = 33;
        final TestStructure[] compoundArray = new TestStructure[numElements];
        for (byte i = 0; i < numElements; ++i)
        {
            final TestStructure testStructure = new TestStructure(i, "Name" + i);
            compoundArray[i] = testStructure;
        }
        VariableArray variableArray = new VariableArray(BigInteger.valueOf(numElements + 1), compoundArray);

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        assertThrows(ZserioError.class, () -> variableArray.write(writer));
        writer.close();
    }

    private BitBuffer writeVariableArrayToBitBuffer(byte numElements) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            writer.writeByte(numElements);
            for (byte i = 0; i < numElements; ++i)
            {
                writer.writeUnsignedInt(i);
                writer.writeString("Name" + i);
            }

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private static final String BLOB_NAME = "variable_array_struct_cast_varuint.blob";
}
